### PR TITLE
feat(hls): Support AES-128 in HLS

### LINKED ISSUE #850: Support AES-128 on HLS
This it's a request feature for support AES-128 on HLS. Is it possible support it?

Sample stream:
https://playertest.longtailvideo.com/adaptive/oceans_aes/oceans_aes.m3u8

Now, I get: HLS_REQUIRED_ATTRIBUTE_MISSING (KEYFORMAT)

### PR DESCRIPTION
**Pull request by theodab**
Expands on the original PR (#3880) by adding support for MP4 and key rotation.

Close #850

Co-authored-by: wjywbs <wjywbs@users.noreply.github.com>

### CODE REVIEW

**Comment by avelad** [comment]
I've seen a bit of PR from the mobile, and I think it would be good to add support for SAMPLE-AES in this same PR since the implementation is similar. what do you think?

File: lib/hls/hls_parser.js [hunk 1]
```diff
@@ -2214,6 +2293,15 @@ shaka.hls.HlsParser = class {
           (i == 0) ? firstStartTime : previousReference.endTime;
       position = mediaSequenceNumber + skippedSegments + i;
 
+      // Apply new AES-128 tags as you see them, keeping a running total.
+      for (const drmTag of item.tags) {
+        if (drmTag.name == 'EXT-X-KEY' &&
+            drmTag.getRequiredAttrValue('METHOD') == 'AES-128') {
+          // eslint-disable-next-line no-await-in-loop
```
[the remark this answers is not in the document]
Comment (theodab) [reply lib/hls/hls_parser.js:2300]: Ah, you're right. This was an async function at first, then when I added the "download keys later" functionality it was no longer async. Thanks for pointing that out!

File: lib/hls/hls_parser.js [hunk 2]
```diff
@@ -1765,6 +1761,82 @@ shaka.hls.HlsParser = class {
...
+    }
+
+    const keyUri = shaka.hls.Utils.constructAbsoluteUri(
+        playlist.absoluteUri, drmTag.getRequiredAttrValue('URI'));
+
+    const requestType = shaka.net.NetworkingEngine.RequestType.KEY;
+    const request = shaka.net.NetworkingEngine.makeRequest(
+        [keyUri], this.config_.retryParameters);
+
+    const keyInfo = /* @type {!shaka.extern.HlsAes128Key} */ ({
```
Comment (joeyparrish) [inline lib/hls/hls_parser.js:1813]: There's a subtle difference here between a cast and a type declaration.

If you annotate `keyInfo` like this:

```js
/* @type {!shaka.extern.HlsAes128Key} */
const keyInfo = {
  // ...
};
```

Then `keyInfo` already has to conform to the record type.  If it doesn't match, it will be fail to compile.

But since you're casting it, you're saying to the compiler "this is definitely HlsAes128Key, ignore what you thought it was".  If we add or change fields later, the compiler will not throw an error.

File: lib/hls/hls_parser.js [hunk 3]
```diff
@@ -1765,6 +1761,82 @@ shaka.hls.HlsParser = class {
...
+      firstMediaSequenceNumber = shaka.hls.Utils.getFirstTagWithNameAsNumber(
+          playlist.tags, 'EXT-X-MEDIA-SEQUENCE', 0);
+    } else {
+      // Exclude 0x at the start of string.
+      iv = shaka.util.Uint8ArrayUtils.fromHex(ivHex.substr(2));
+      if (iv.byteLength != 16) {
+        throw new shaka.util.Error(
+            shaka.util.Error.Severity.CRITICAL,
+            shaka.util.Error.Category.MANIFEST,
+            shaka.util.Error.Code.HLS_AES_128_INVALID_KEY_OR_IV_LENGTH,
```
Comment (joeyparrish) [inline lib/hls/hls_parser.js:1801]: I would like you to split this into one error for key length and another for IV length.  It is useful when analyzing error logs for people to be able to point to one cause and not multiple possible causes.

File: lib/media/segment_reference.js [hunk 4]
```diff
@@ -166,12 +166,15 @@ shaka.media.SegmentReference = class {
...
+   * @param {shaka.extern.HlsAes128Key=} hlsAes128Key
+   *  The segment's AES-128-CBC full segment encryption key and iv.
    */
   constructor(
       startTime, endTime, uris, startByte, endByte, initSegmentReference,
       timestampOffset, appendWindowStart, appendWindowEnd,
       partialReferences = [], tilesLayout = '', tileDuration = null,
-      syncTime = null, status = shaka.media.SegmentReference.Status.AVAILABLE) {
+      syncTime = null, status = shaka.media.SegmentReference.Status.AVAILABLE,
+      hlsAes128Key) {
```
Comment (joeyparrish) [inline lib/media/segment_reference.js:177]: I think if the parameters before this have defaults, this should, too.

File: lib/media/segment_reference.js [hunk 5]
```diff
@@ -233,6 +236,9 @@ shaka.media.SegmentReference = class {
 
     /** @type {shaka.media.SegmentReference.Status} */
     this.status = status;
+
+    /** @type {shaka.extern.HlsAes128Key|undefined} */
```
Comment (joeyparrish) [inline lib/media/segment_reference.js:240]: If you default to null, this would be nullable instead of |undefined.

File: lib/media/streaming_engine.js [hunk 6]
```diff
@@ -1273,11 +1276,15 @@ shaka.media.StreamingEngine = class {
             'ReadableStream is not supported by the browser.');
         }
         const fetchSegment = this.fetch_(mediaState, reference);
-        const result = await fetchSegment;
+        let result = await fetchSegment;
         this.destroyer_.ensureNotDestroyed();
         if (this.fatalError_) {
           return;
         }
+        if (reference.hlsAes128Key && iter) {
```
Comment (joeyparrish) [inline lib/media/streaming_engine.js:1284]: When is the iterator null?  A few lines down on 1310 we assume it's not.

If this is to satisfy the compiler that a nullable thing is non-null, I think an assertion would be preferable (for clarity).  This way, it looks like this is a condition we need to handle by skipping decryption, which doesn't make sense to me.

File: lib/media/streaming_engine.js [hunk 7]
```diff
@@ -1368,6 +1375,34 @@ shaka.media.StreamingEngine = class {
...
   }
 
+  /**
+   * @param {!BufferSource} rawResult
+   * @param {!shaka.media.SegmentReference} reference
+   * @param {!shaka.media.SegmentIterator} iter
+   * @return {!Promise.<!BufferSource>} finalResult
+   * @private
+   */
+  async aes128Decrypt_(rawResult, reference, iter) {
```
Comment (joeyparrish) [inline lib/media/streaming_engine.js:1385]: It would be nice to have tests that covered this somehow.  I know StreamingEngine tests are a mess, but could you give it a shot?  If it can't be done in a day, then please file an issue under "type: code health" to follow up later.

**Review by joeyparrish — CHANGES_REQUESTED** [review]

File: externs/shaka/manifest.js [hunk 8]
```diff
@@ -244,6 +244,39 @@ shaka.extern.Variant;
...
+ *
+ * @description
+ * AES-128 key and iv info from the HLS manifest.
+ *
+ * @property {string} method
+ *   The key method defined in the HLS manifest.
+ * @property {webCrypto.CryptoKey|undefined} cryptoKey
+ *   Web crypto key object of the AES-128 CBC key. If unset, the "fetchKey"
+ *   property should be provided.
+ * @property {shaka.extern.CreateSegmentIndexFunction|undefined} fetchKey
```
[the remark this answers is not in the document]
Comment (theodab) [reply externs/shaka/manifest.js]: Done.

File: package-lock.json [hunk 9]
```diff
@@ -2975,14 +2975,20 @@
       }
     },
     "node_modules/caniuse-lite": {
-      "version": "1.0.30001312",
```
Comment (avelad) [inline package-lock.json:2978]: @theodab Is this change necessary in this PR?

Comment (theodab) [reply package-lock.json]: Ah, no, that was an accident.

Comment (joeyparrish) [reply package-lock.json]: The difference between "npm install" and "npm ci" is that "install" will update the package lock and find new versions that satisfy the requirements in package.json.  "ci" will only install exactly what is in the lock file, for consistent reproducibility in CI systems.

**Review by joeyparrish — APPROVED** [review]
LGTM!
